source/scpp/extra/DVMChecks.cpp

```diff
@@ -0,0 +1,320 @@
+/*******************************************************************************
+
+    Contains checking the offset of virtual methods.
+
+    Note:
+        This was automatically generated by agora-gen.
+        Do not modify this file.
+        If you want to modify it, please use agora-gen.
+        agora-gen source/scpp/extra/DVMChecks.cpp
+        or
+        dub -c gen -- source/scpp/extra/DVMChecks.cpp
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+#include "DVMChecks.h"
+
+///  Uses to test TestA.
+class TestAForTest
+{
+    public:
+        virtual ~TestAForTest () {};
+        virtual void vfunc1 () {};
+        virtual void vfunc2 () {};
+};
+
+///  Checks the offset of virtual methods inside the class TestA.
+int checkVMOffsetTestA ()
+{
+    auto df1 = &TestAForTest::vfunc1;
+    auto cf1 = &TestA::vfunc1;
+    if (reinterpret_cast<void *&>(df1) != reinterpret_cast<void *&>(cf1))
+        return 1;
+
+    auto df2 = &TestAForTest::vfunc2;
+    auto cf2 = &TestA::vfunc2;
+    if (reinterpret_cast<void *&>(df2) != reinterpret_cast<void *&>(cf2))
+        return 2;
+
+    return 0;
+}
+
+///  Returns the offset of virtual methods inside the class TestA.
+long getVMOffsetTestA (const char* name)
+{
+    if (strcmp(name, "vfunc1") == 0)
+    {
+        auto pf = &TestA::vfunc1;
+        return ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
+    }
+    else if (strcmp(name, "vfunc2") == 0)
+    {
+        auto pf = &TestA::vfunc2;
+        return ((long)reinterpret_cast<void *&>(pf)) / sizeof(long);
+    }
+    else
+    {
+        return -1;
+    }
+}
+
+///  Uses to test SCPDriver.
+class SCPDriverForTest
```

Comment:
We shouldn't need to declare this extra type on the C++ side. Can't we just get the offset on the D side ?

Reply:
> We shouldn't need to declare this extra type on the C++ side. Can't we just get the offset on the D side ?
I wonder if this type of data is possible.
map<string, int>